src/3dDiff.c

```diff
@@ -0,0 +1,407 @@
+            quiet = 1;
+            iarg++; continue;
+        }
+
+        /* OPTIONAL: long_report */
+        if ( strcmp(argv[iarg],"-long_report") == 0) {
+            long_report = 1;
+            iarg++; continue;
+        }
+
+        /* left image */
+        if ( strncmp(argv[iarg],"-a",2) == 0) {
+            if (++iarg >= argc) diff3d_crash("Need dset after -a");
+            a_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+
+        /* right image */
+        if ( strncmp(argv[iarg],"-b",2) == 0) {
+            if (++iarg >= argc) diff3d_crash("Need dset after -b");
+            b_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+        
+        ERROR_message("ILLEGAL option: %s\n", argv[iarg]) ;
+        suggest_best_prog_option(argv[0], argv[iarg]);
+        exit(1);
+    }
+
+    /* Validate args */
+    if ( !a_fname )
+        diff3d_crash("No dset a supplied!");
+    if ( !b_fname )
+        diff3d_crash("No dset b supplied!");
+    /* tolerance is guaranteed to exist, validation performed above */
+
+    /*   Thanks to PT for simplifying options for me */
+    disp_opt_sum = brutalist + tabular + quiet + long_report;
+    if ( !disp_opt_sum ){
+        report = 1;
+    }
+    else if ( disp_opt_sum > 1 ){
+       diff3d_crash("Must choose ONLY one of these display opts:"
+                  "  -brutalist, -tabular, -q, -long_report");
+    }
+    // ... and in any other case, one should have exactly 1 option
+    // entered by the user, and be OK
+
+
+    /* Load the images and check for mutual compatibility*/
+    set_obliquity_report(0); /* We'll check that ourselves below */
+    ds1 = THD_open_dataset(a_fname);
+    if ( ds1 == NULL ) diff3d_crash("Can't open dataset %s", a_fname);
+    if ( mask_fname ) DSET_mallocize(ds1);
+    DSET_load(ds1); CHECK_LOAD_ERROR(ds1);
+    ds2 = THD_open_dataset(b_fname);
+    if ( ds2 == NULL ) diff3d_crash("Can't open dataset %s", b_fname);
+    if ( mask_fname ) DSET_mallocize(ds2);
+    DSET_load(ds2); CHECK_LOAD_ERROR(ds2);
+    if ( THD_dataset_mismatch( ds1, ds2 ) ) {
+        diff3d_crash("Mismatch between dsets!\n");
+    }
+    if ( DSET_NVALS(ds1) != DSET_NVALS(ds2) ) {
+        diff3d_crash(
+            "Incompatible time points: dset %s contains %d timepoints, dset %s contains %d timepoints",
+            a_fname, DSET_NVALS(ds1), b_fname, DSET_NVALS(ds2)
+        );
+    }
+    nt = DSET_NVALS(ds1);
+    nv = DSET_NVOX(ds1);
+    counts = (int*) calloc(nv, sizeof(int));
+    /* If there's a mask, validate and apply it to the dsets */
+    if ( mask_fname ) {
+        total_elements = 0;
+        THD_3dim_dataset * ds_mask = THD_open_dataset(mask_fname);
+        DSET_load(ds_mask); CHECK_LOAD_ERROR(ds_mask);
+        if ( THD_dataset_mismatch( ds1, ds_mask ) ) {
+            diff3d_crash("Mismatch between input and mask dsets!\n");
+        }
+        if ( DSET_NVOX(ds_mask) != nv ) {
+            diff3d_crash("WTF bro?");
```

Comment:
How about slightly different message here?

Reply:
That isn't actually even necessary because I check for a mismatch above, it was during angry debugging time and I forgot to remove it.

Reply:
OK. And note that often (or always?) a mask will just have 1 volume, even when applied to a multi-volume dataset.

Reply:
Always, in this case, because I just select the first sub-brick. I can note that in the help.